### PR TITLE
[3.10] gh-102950: Implement PEP 706 – Filter for tarfile.extractall (GH-102953)

### PR DESCRIPTION
**Pull request by mcepl**
Manual port of gh-102953 to 3.10.

Please, review.

* Issue: gh-102950

### CODE REVIEW

**Comment by encukou** [comment]
Thanks for getting to this sooner than I did!
But see https://github.com/python/cpython/pull/103832/commits for the changes needed in the backport.

**Comment by mcepl** [comment]
I was basing it on this 3.11 PR, I know better than using master branch for this. But yes, I will re-check, that everything in that PR has been included.

**Comment by encukou** [comment]
Ah, sorry! I just looked at the `versionchanged:: 3.12` and gave a quick comment. Will do a proper review later.

**Comment by mcepl** [comment]
Yes, you were right, I missed those, not sure how I have managed to do that.

**Comment by mcepl** [comment]
I don't understand those Windows failures … shouldn’t it be covered by https://github.com/python/cpython/pull/104128/commits/ae62b3a3e3dc48b58480e1834b83bdd3b7365b0c ?

File: Doc/library/shutil.rst [hunk 1]
```diff
@@ -634,8 +634,9 @@ provided.  They rely on the :mod:`zipfile` and :mod:`tarfile` modules.
    registered for that extension.  In case none is found,
    a :exc:`ValueError` is raised.
 
-   The keyword-only *filter* argument is passed to the underlying unpacking
-   function. For zip files, *filter* is not accepted.
+   The keyword-only *filter* argument, which was added in Python 3.11.4,
```
Comment (encukou) [inline Doc/library/shutil.rst]: This should be 3.10.12 everywhere.

**Comment by encukou** [comment]
No. ae62b3a3e3dc48b58480e1834b83bdd3b7365b0c is for Solaris or BSD, supported platforms (per PEP 11) don't need it.

This change made me get a Windows VM, so I'll investigate there. can't promise it'll be this week though.

**Comment by encukou** [comment]
It's related to a change made while backporting, going from
```python
        if mode is not None and os_helper.can_chmod():
```
to just
```python
        if mode is not None:
```

The `can_chmod()` determines whether chmod can set all the mode bits. According to [the docs](https://docs.python.org/3/library/os.html), it's only limited on Windows, so I'll hardcode that. (It's also limited on WASI, but in 3.10 tests we don't care.)

I can't push fixes to your repo so I opened #104327.

**Comment by mcepl** [comment]
> I can't push fixes to your repo so I opened #104327.

Sorry about that, I don’t know how to set it up with the existing PR.

**Comment by mcepl** [comment]
Couldn’t you just make PR to https://github.com/openSUSE-Python/cpython/tree/PEP706-for-py310 ?

**Comment by encukou** [comment]
I don't think I could run CPython buildbots on another repo.
But since you've updated this PR quickly, I'll run them here.

**Comment by encukou** [comment]
The red checks are unrelated: Python 3.10 doesn't build on Wasm, and RtD's OpenSSL is too old for urllib3.

**Comment by encukou** [comment]
Thanks for the backport!

**Comment by mcepl** [comment]
So, now 3.9! :)

**Pull request merged**